src/wasm/wasm-type.cpp

```diff
@@ -341,11 +363,67 @@ TypeInfo::~TypeInfo() {
   WASM_UNREACHABLE("unexpected kind");
 }
 
+bool TypeInfo::getCanonical(Type& out) const {
+  if (isTuple()) {
+    if (tuple.types.size() == 0) {
+      out = Type::none;
+      return true;
+    }
+    if (tuple.types.size() == 1) {
+      out = tuple.types[0];
+      return true;
```

Comment:
still trying to reason about what this method does, I got stuck here. It looks like if a tuple of size 1 arrives here then we return that single value, and `true`. Why is the "canonical" value of the tuple that type? Maybe this is using "canonical" in a way I don't follow.

Reply:
This code is almost exactly what `TypeStore::canonicalize` used to do, if that helps.

Reply:
For the tuple case specifically: we don't allow singleton tuple types, so if you try to construct one, you should get the single type back instead of a fresh type.

Reply:
Should we make it an error to try to construct a tuple of size 1 then?

Reply:
I don't think so. Code that currently constructs types that may or may not be tuples can currently treat both cases uniformly, but if we made it an error to even try to construct a tuple of size 1 (or zero), that code would have to start dealing with the new discontinuity explicitly.